stlab/forest_algorithms.hpp

```diff
@@ -0,0 +1,141 @@
+/**************************************************************************************************/
+
+#ifndef STLAB_FOREST_ALGORITHMS_HPP
+#define STLAB_FOREST_ALGORITHMS_HPP
+
+/**************************************************************************************************/
+
+// stlab
+#include <stlab/concurrency/optional.hpp>
+#include <stlab/forest.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+
+namespace forests {
+
+/**************************************************************************************************/
+// "Congruent" would be a nice name here, but in geometry that also implies reflection.
+template <class Forest1, class Forest2>
+bool equal_shape(const Forest1& x, const Forest2& y) {
+    if (x.size_valid() && y.size_valid() && x.size() != y.size()) return false;
+    auto pos{y.begin()};
+    for (auto first(x.begin()), last(x.end()); first != last; ++first, ++pos) {
+        if (first.edge() != pos.edge()) return false;
+    }
+    return true;
+}
+
+/**************************************************************************************************/
+
+template <class Container>
+struct insert_iterator {
```

Comment:
@sean-parent this is the more fleshed out `insert_iterator`. Unlike `std::insert_iterator`, `operator++` cannot be a no-op, and `operator=` cannot increment the iterator immediately after insertion. It also needs a `trailing()` function to pivot the iterator.

These requirements are unique to the `transform` algorithm below, and allow the output iterator to replicate the shape of the source forest during output. If a user has no need of these requirements, the standard library algorithms will work just fine over a `preorder_range`.

Reply:
Maybe make this a `detail` of transcribe - or call it a `transcribe_iterator`. It is something more than an OutputIterator and less than a ForwardIterator - and certainly doesn't match `std::insert_iterator` in semantics.

Reply:
I'll call it a `transcribe_iterator` - users will instantiate these things in their own code so putting it in the `detail` namespace doesn't make sense.

Reply:
👍